src/tty/windows.rs

```diff
@@ -225,13 +233,78 @@ impl RawReader for ConsoleRawReader {
+                            KeyEvent(K::Char(ch @ '['), M::NONE),
+                        )
+                        | (['\x1b', '[', '\0', '\0', '\0'], KeyEvent(K::Char(ch @ '2'), M::NONE))
+                        | (['\x1b', '[', '2', '\0', '\0'], KeyEvent(K::Char(ch @ '0'), M::NONE))
+                        | (['\x1b', '[', '2', '0', '\0'], KeyEvent(K::Char(ch @ '0'), M::NONE))
+                        | (['\x1b', '[', '2', '0', '\0'], KeyEvent(K::Char(ch @ '1'), M::NONE)) => {
+                            esc_seq[esc_seq_len] = ch;
+                            esc_seq_len += 1;
+                            continue;
+                        }
+                        (['\x1b', '[', '2', '0', '0'], KeyEvent(K::Char('~'), M::NONE)) => {
+                            debug!(target: "rustyline", "Bracketed paste start");
+                            key = KeyEvent(K::BracketedPasteStart, M::NONE);
+                        }
+                        (['\x1b', '[', '2', '0', '1'], KeyEvent(K::Char('~'), M::NONE)) => {
+                            debug!(target: "rustyline", "Bracketed paste end");
+                            key = KeyEvent(K::BracketedPasteEnd, M::NONE);
+                        }
+                        (_, KeyEvent(K::Char(ch), M::NONE)) => {
+                            debug!(target: "rustyline", "unsupported esc sequence: \\E{}{}", esc_seq[1..esc_seq_len].iter().cloned().collect::<String>(), ch);
+                            key = KeyEvent(K::UnknownEscSeq, M::NONE);
+                        }
+                        _ => {
+                            debug!(target: "rustyline", "unsupported esc sequence: \\E{}", esc_seq[1..esc_seq_len].iter().cloned().collect::<String>());
+                            key = KeyEvent(K::UnknownEscSeq, M::NONE);
+                        }
+                    }
+                }
+            }
+
             return Ok(key);
         }
     }
 
     fn read_pasted_text(&mut self) -> Result<String> {
-        Ok(clipboard_win::get_clipboard_string()?)
+        if self.enable_bracketed_paste {
+            let mut buffer = String::new();
+
+            loop {
+                match self.next_key(true)? {
+                    KeyEvent(K::BracketedPasteEnd, _) => {
+                        let buffer = buffer.replace("\r\n", "\n");
+                        let buffer = buffer.replace('\r', "\n");
+                        return Ok(buffer);
+                    }
+                    KeyEvent(K::Char(ch), M::NONE) => {
+                        buffer.push(ch);
+                    }
+                    KeyEvent(K::Char('M'), M::CTRL) => {
+                        buffer.push('\r');
+                    }
+                    KeyEvent(K::Char('J'), M::CTRL) => {
+                        buffer.push('\n');
+                    }
+                    _ => (),
+                }
+            }
+        } else {
+            Ok(clipboard_win::get_clipboard_string()?)
```

Comment:
What about a custom binding that should paste ?

Reply:
I suppose, with bracketed paste on, it pasting directly from clipboard should be disabled?

Reply:
So maybe on Windows bracketed paste should default to `false`?

Reply:
Set `enable_bracketed_paste` to default to `false` on Windows.